{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using DESeq2 for sample normalization and statistical analysis to identify mRNA associated with smoke exposure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code includes: data organization,PCA, surrogate variable analysis, and other steps prior to DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in install.packages(\"tibble\"):\n",
      "“installation of package ‘tibble’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Warning message in install.packages(\"rlang\"):\n",
      "“installation of package ‘rlang’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Warning message in install.packages(\"RcppArmadillo\"):\n",
      "“installation of package ‘RcppArmadillo’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Bioconductor version 3.11 (BiocManager 1.30.10), R 4.0.2 (2020-06-22)\n",
      "\n",
      "Installing package(s) 'DESeq2'\n",
      "\n",
      "also installing the dependencies ‘bit’, ‘bitops’, ‘bit64’, ‘plogr’, ‘RCurl’, ‘GenomeInfoDbData’, ‘zlibbioc’, ‘lambda.r’, ‘futile.options’, ‘RSQLite’, ‘XML’, ‘GenomeInfoDb’, ‘XVector’, ‘DelayedArray’, ‘futile.logger’, ‘snow’, ‘BH’, ‘AnnotationDbi’, ‘annotate’, ‘S4Vectors’, ‘IRanges’, ‘GenomicRanges’, ‘SummarizedExperiment’, ‘BiocGenerics’, ‘Biobase’, ‘BiocParallel’, ‘genefilter’, ‘locfit’, ‘geneplotter’\n",
      "\n",
      "\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘bit’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘bitops’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘zlibbioc’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘XML’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘locfit’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘bit64’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘RCurl’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘S4Vectors’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Biobase’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘RSQLite’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘IRanges’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘BiocParallel’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘GenomeInfoDb’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘XVector’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘DelayedArray’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘AnnotationDbi’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘annotate’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘GenomicRanges’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘SummarizedExperiment’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘genefilter’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘geneplotter’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘DESeq2’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'SQUAREM', 'broom', 'callr', 'cli', 'clipr', 'data.table',\n",
      "  'digest', 'foreach', 'htmlwidgets', 'iterators', 'labeling', 'lme4',\n",
      "  'pbdZMQ', 'ps', 'readr', 'recipes', 'rlang', 'rmarkdown', 'sp', 'tibble'\n",
      "\n",
      "Bioconductor version 3.11 (BiocManager 1.30.10), R 4.0.2 (2020-06-22)\n",
      "\n",
      "Installing package(s) 'sva'\n",
      "\n",
      "also installing the dependencies ‘bit’, ‘bit64’, ‘bitops’, ‘IRanges’, ‘RSQLite’, ‘S4Vectors’, ‘XML’, ‘RCurl’, ‘AnnotationDbi’, ‘annotate’, ‘Biobase’, ‘locfit’, ‘genefilter’, ‘BiocParallel’, ‘limma’, ‘edgeR’\n",
      "\n",
      "\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘bit’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘bitops’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘S4Vectors’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘XML’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Biobase’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘locfit’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘BiocParallel’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘limma’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘bit64’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘IRanges’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘RCurl’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘edgeR’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘RSQLite’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘AnnotationDbi’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘annotate’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘genefilter’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘sva’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'SQUAREM', 'broom', 'callr', 'cli', 'clipr', 'data.table',\n",
      "  'digest', 'foreach', 'htmlwidgets', 'iterators', 'labeling', 'lme4',\n",
      "  'pbdZMQ', 'ps', 'readr', 'recipes', 'rlang', 'rmarkdown', 'sp', 'tibble'\n",
      "\n",
      "Bioconductor version 3.11 (BiocManager 1.30.10), R 4.0.2 (2020-06-22)\n",
      "\n",
      "Installing package(s) 'BiocManager'\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'SQUAREM', 'broom', 'callr', 'cli', 'clipr', 'data.table',\n",
      "  'digest', 'foreach', 'htmlwidgets', 'iterators', 'labeling', 'lme4',\n",
      "  'pbdZMQ', 'ps', 'readr', 'recipes', 'rlang', 'rmarkdown', 'sp', 'tibble'\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "also installing the dependencies ‘bitops’, ‘gtools’, ‘caTools’\n",
      "\n",
      "\n",
      "Warning message in install.packages(\"gplots\"):\n",
      "“installation of package ‘bitops’ had non-zero exit status”\n",
      "Warning message in install.packages(\"gplots\"):\n",
      "“installation of package ‘gtools’ had non-zero exit status”\n",
      "Warning message in install.packages(\"gplots\"):\n",
      "“installation of package ‘caTools’ had non-zero exit status”\n",
      "Warning message in install.packages(\"gplots\"):\n",
      "“installation of package ‘gplots’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Warning message in install.packages(\"data.table\"):\n",
      "“installation of package ‘data.table’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Installing DESeq2:\n",
    "# Note that BiocManager allows you to access the Bioconductor repo in which DeSeq2 package is. DeSeq2 pacakge has a \"tibble\" \"RcppArmadillo\" and \"rlang\" dependency.\n",
    "# to install RcppArmadillo, you must install a Fortan compiler (I used gfortran-6.1.pkg, https://cran.r-project.org/bin/macosx/tools/).\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "\n",
    "install.packages(\"tibble\")\n",
    "install.packages(\"rlang\")\n",
    "install.packages(\"RcppArmadillo\")\n",
    "\n",
    "# Install DESeq and SVA:\n",
    "BiocManager::install(\"DESeq2\", version = \"3.11\")\n",
    "BiocManager::install(\"sva\", version =\"3.11\")\n",
    "BiocManager::install(\"BiocManager\")\n",
    "\n",
    "# Install plot packages\n",
    "install.packages(\"ggbeeswarm\") # note that I had to manually install this package using tools --> install packages\n",
    "#install.packages(\"gridExtra\")\n",
    "\n",
    "# Install tidyverse packages \n",
    "#install.packages(\"tidyverse\")\n",
    "\n",
    "# Install gplots packages\n",
    "install.packages(\"gplots\")\n",
    "\n",
    "# Install RColorBrewer package\n",
    "install.packages(\"RColorBrewer\")\n",
    "\n",
    "# Install scales (needed for tidyverse activation):\n",
    "install.packages(\"scales\")\n",
    "\n",
    "# Install data.tables\n",
    "install.packages(\"data.table\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Activating appropriate libraries (skip step above if everything is already installed)\n",
    "library(data.table)\n",
    "library(rlang)\n",
    "library(DESeq2)\n",
    "library(limma)\n",
    "library(sva)\n",
    "library(ggplot2)\n",
    "library(stats)\n",
    "library(scales)\n",
    "library(ggbeeswarm)\n",
    "library(gridExtra)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(gplots)\n",
    "library(RColorBrewer)\n",
    "library(data.table)\n",
    "library(readxl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory\n",
    "setwd('/Users/alexis/IEHS Dropbox/Rager Lab/Alexis_Payton/2_Wildfire_Analysis/Input')\n",
    "\n",
    "# Create an output folder (make sure to make the folder first, and then point to it here)\n",
    "Output <- (\"/Users/alexis/IEHS Dropbox/Rager Lab/Alexis_Payton/2_Wildfire_Analysis/Output\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading count data and metadata (subject information) & organizing and filtering to keep subjects present in both files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.0 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.0.3     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.3.1     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.0\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32m%@%()\u001b[39m         masks \u001b[34mrlang\u001b[39m::%@%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mrlang\u001b[39m::\u001b[32m:=()\u001b[39m          masks \u001b[34mdata.table\u001b[39m:::=()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mas_function()\u001b[39m masks \u001b[34mrlang\u001b[39m::as_function()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m      masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m       masks \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten()\u001b[39m     masks \u001b[34mrlang\u001b[39m::flatten()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten_chr()\u001b[39m masks \u001b[34mrlang\u001b[39m::flatten_chr()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten_dbl()\u001b[39m masks \u001b[34mrlang\u001b[39m::flatten_dbl()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten_int()\u001b[39m masks \u001b[34mrlang\u001b[39m::flatten_int()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten_lgl()\u001b[39m masks \u001b[34mrlang\u001b[39m::flatten_lgl()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten_raw()\u001b[39m masks \u001b[34mrlang\u001b[39m::flatten_raw()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32minvoke()\u001b[39m      masks \u001b[34mrlang\u001b[39m::invoke()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m         masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m        masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mlist_along()\u001b[39m  masks \u001b[34mrlang\u001b[39m::list_along()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mmodify()\u001b[39m      masks \u001b[34mrlang\u001b[39m::modify()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mprepend()\u001b[39m     masks \u001b[34mrlang\u001b[39m::prepend()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32msplice()\u001b[39m      masks \u001b[34mrlang\u001b[39m::splice()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse) #remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(readxl) #remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Missing column names filled in: 'X1' [1]”\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\n",
      "See spec(...) for full column specifications.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Read in count data\n",
    "# countdata <- read.table(file = 'Wildfire_GeneCounts_102320.csv', header = TRUE)\n",
    "countdata <- read_csv(file = 'Wildfire_GeneCounts_102320.csv') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: Problem with `filter()` input `..1`.\n\u001b[31m✖\u001b[39m Input `..1` must be of size 30146 or 1, not size 5.\n\u001b[34mℹ\u001b[39m Input `..1` is ``%in%`(...)`.\n",
     "output_type": "error",
     "traceback": [
      "Error: Problem with `filter()` input `..1`.\n\u001b[31m✖\u001b[39m Input `..1` must be of size 30146 or 1, not size 5.\n\u001b[34mℹ\u001b[39m Input `..1` is ``%in%`(...)`.\nTraceback:\n",
      "1. countdata %>% filter(colnames(countdata)[1:5] %in% c(\"MH112\", \n .     \"MH113\", \"MH114\", \"MH115\", \"MH116\", \"MH117\", \"MH118\", \"MH101\", \n .     \"MH102\", \"MH103\", \"MH104\", \"MH105\", \"MH106\"))",
      "2. withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))",
      "3. eval(quote(`_fseq`(`_lhs`)), env, env)",
      "4. eval(quote(`_fseq`(`_lhs`)), env, env)",
      "5. `_fseq`(`_lhs`)",
      "6. freduce(value, `_function_list`)",
      "7. withVisible(function_list[[k]](value))",
      "8. function_list[[k]](value)",
      "9. filter(., colnames(countdata)[1:5] %in% c(\"MH112\", \"MH113\", \"MH114\", \n .     \"MH115\", \"MH116\", \"MH117\", \"MH118\", \"MH101\", \"MH102\", \"MH103\", \n .     \"MH104\", \"MH105\", \"MH106\"))",
      "10. filter.data.frame(., colnames(countdata)[1:5] %in% c(\"MH112\", \n  .     \"MH113\", \"MH114\", \"MH115\", \"MH116\", \"MH117\", \"MH118\", \"MH101\", \n  .     \"MH102\", \"MH103\", \"MH104\", \"MH105\", \"MH106\"))",
      "11. filter_rows(.data, ...)",
      "12. withCallingHandlers(mask$eval_all_filter(dots, env_filter), error = function(e) {\n  .     local_call_step(dots = dots, .index = env_filter$current_expression, \n  .         .fn = \"filter\")\n  .     abort(c(cnd_bullet_header(), x = conditionMessage(e), i = cnd_bullet_input_info(), \n  .         i = cnd_bullet_cur_group_label()), class = \"dplyr_error\")\n  . })",
      "13. mask$eval_all_filter(dots, env_filter)",
      "14. abort_glue(\"Input `..{index}` must be of size {or_1(expected_size)}, not size {size}.\", \n  .     list(index = 1L, size = 5L, expected_size = 30146L), \"dplyr:::filter_incompatible_size\")",
      "15. exec(abort, message = message, class = class, !!!.envir)",
      "16. (function (message = NULL, class = NULL, ..., trace = NULL, parent = NULL, \n  .     .subclass) \n  . {\n  .     validate_signal_args(.subclass)\n  .     if (is_null(trace) && is_null(peek_option(\"rlang:::disable_trace_capture\"))) {\n  .         with_options(`rlang:::disable_trace_capture` = TRUE, \n  .             {\n  .                 trace <- trace_back()\n  .                 if (is_null(parent)) {\n  .                   context <- trace_length(trace)\n  .                 }\n  .                 else {\n  .                   context <- find_capture_context(3L)\n  .                 }\n  .                 trace <- trace_trim_context(trace, context)\n  .             })\n  .     }\n  .     message <- validate_signal_message(message, class)\n  .     message <- collapse_cnd_message(message)\n  .     cnd <- error_cnd(class, ..., message = message, parent = parent, \n  .         trace = trace)\n  .     signal_abort(cnd)\n  . })(message = structure(\"Input `..1` must be of size 30146 or 1, not size 5.\", class = c(\"glue\", \n  . \"character\")), class = \"dplyr:::filter_incompatible_size\", index = 1L, \n  .     size = 5L, expected_size = 30146L)",
      "17. signal_abort(cnd)",
      "18. signalCondition(cnd)",
      "19. (function (e) \n  . {\n  .     local_call_step(dots = dots, .index = env_filter$current_expression, \n  .         .fn = \"filter\")\n  .     abort(c(cnd_bullet_header(), x = conditionMessage(e), i = cnd_bullet_input_info(), \n  .         i = cnd_bullet_cur_group_label()), class = \"dplyr_error\")\n  . })(structure(list(message = structure(\"Input `..1` must be of size 30146 or 1, not size 5.\", class = c(\"glue\", \n  . \"character\")), trace = structure(list(calls = list(IRkernel::main(), \n  .     kernel$run(), IRkernel:::handle_shell(), executor$execute(msg), \n  .     base::tryCatch(evaluate(request$content$code, envir = .GlobalEnv, \n  .         output_handler = oh, stop_on_error = 1L), interrupt = function(cond) {\n  .         log_debug(\"Interrupt during execution\")\n  .         interrupted <<- TRUE\n  .     }, error = .self$handle_error), base:::tryCatchList(expr, \n  .         classes, parentenv, handlers), base:::tryCatchOne(tryCatchList(expr, \n  .         names[-nh], parentenv, handlers[-nh]), names[nh], parentenv, \n  .         handlers[[nh]]), base:::doTryCatch(return(expr), name, \n  .         parentenv, handler), base:::tryCatchList(expr, names[-nh], \n  .         parentenv, handlers[-nh]), base:::tryCatchOne(expr, names, \n  .         parentenv, handlers[[1L]]), base:::doTryCatch(return(expr), \n  .         name, parentenv, handler), evaluate::evaluate(request$content$code, \n  .         envir = .GlobalEnv, output_handler = oh, stop_on_error = 1L), \n  .     evaluate:::evaluate_call(expr, parsed$src[[i]], envir = envir, \n  .         enclos = enclos, debug = debug, last = i == length(out), \n  .         use_try = stop_on_error != 2L, keep_warning = keep_warning, \n  .         keep_message = keep_message, output_handler = output_handler, \n  .         include_timing = include_timing), evaluate:::timing_fn(handle(ev <- withCallingHandlers(withVisible(eval(expr, \n  .         envir, enclos)), warning = wHandler, error = eHandler, \n  .         message = mHandler))), evaluate:::handle(ev <- withCallingHandlers(withVisible(eval(expr, \n  .         envir, enclos)), warning = wHandler, error = eHandler, \n  .         message = mHandler)), base::try(f, silent = TRUE), base::tryCatch(expr, \n  .         error = function(e) {\n  .             call <- conditionCall(e)\n  .             if (!is.null(call)) {\n  .                 if (identical(call[[1L]], quote(doTryCatch))) \n  .                   call <- sys.call(-4L)\n  .                 dcall <- deparse(call)[1L]\n  .                 prefix <- paste(\"Error in\", dcall, \": \")\n  .                 LONG <- 75L\n  .                 sm <- strsplit(conditionMessage(e), \"\\n\")[[1L]]\n  .                 w <- 14L + nchar(dcall, type = \"w\") + nchar(sm[1L], \n  .                   type = \"w\")\n  .                 if (is.na(w)) \n  .                   w <- 14L + nchar(dcall, type = \"b\") + nchar(sm[1L], \n  .                     type = \"b\")\n  .                 if (w > LONG) \n  .                   prefix <- paste0(prefix, \"\\n  \")\n  .             }\n  .             else prefix <- \"Error : \"\n  .             msg <- paste0(prefix, conditionMessage(e), \"\\n\")\n  .             .Internal(seterrmessage(msg[1L]))\n  .             if (!silent && isTRUE(getOption(\"show.error.messages\"))) {\n  .                 cat(msg, file = outFile)\n  .                 .Internal(printDeferredWarnings())\n  .             }\n  .             invisible(structure(msg, class = \"try-error\", condition = e))\n  .         }), base:::tryCatchList(expr, classes, parentenv, handlers), \n  .     base:::tryCatchOne(expr, names, parentenv, handlers[[1L]]), \n  .     base:::doTryCatch(return(expr), name, parentenv, handler), \n  .     base::withCallingHandlers(withVisible(eval(expr, envir, enclos)), \n  .         warning = wHandler, error = eHandler, message = mHandler), \n  .     base::withVisible(eval(expr, envir, enclos)), base::eval(expr, \n  .         envir, enclos), base::eval(expr, envir, enclos), countdata %>% \n  .         filter(colnames(countdata)[1:5] %in% c(\"MH112\", \"MH113\", \n  .             \"MH114\", \"MH115\", \"MH116\", \"MH117\", \"MH118\", \"MH101\", \n  .             \"MH102\", \"MH103\", \"MH104\", \"MH105\", \"MH106\")), base::withVisible(eval(quote(`_fseq`(`_lhs`)), \n  .         env, env)), base::eval(quote(`_fseq`(`_lhs`)), env, env), \n  .     base::eval(quote(`_fseq`(`_lhs`)), env, env), `_fseq`(`_lhs`), \n  .     magrittr::freduce(value, `_function_list`), base::withVisible(function_list[[k]](value)), \n  .     function_list[[k]](value), dplyr::filter(., colnames(countdata)[1:5] %in% \n  .         c(\"MH112\", \"MH113\", \"MH114\", \"MH115\", \"MH116\", \"MH117\", \n  .             \"MH118\", \"MH101\", \"MH102\", \"MH103\", \"MH104\", \"MH105\", \n  .             \"MH106\")), dplyr:::filter.data.frame(., colnames(countdata)[1:5] %in% \n  .         c(\"MH112\", \"MH113\", \"MH114\", \"MH115\", \"MH116\", \"MH117\", \n  .             \"MH118\", \"MH101\", \"MH102\", \"MH103\", \"MH104\", \"MH105\", \n  .             \"MH106\")), dplyr:::filter_rows(.data, ...), base::withCallingHandlers(mask$eval_all_filter(dots, \n  .         env_filter), error = function(e) {\n  .         local_call_step(dots = dots, .index = env_filter$current_expression, \n  .             .fn = \"filter\")\n  .         abort(c(cnd_bullet_header(), x = conditionMessage(e), \n  .             i = cnd_bullet_input_info(), i = cnd_bullet_cur_group_label()), \n  .             class = \"dplyr_error\")\n  .     }), mask$eval_all_filter(dots, env_filter), dplyr:::abort_glue(\"Input `..{index}` must be of size {or_1(expected_size)}, not size {size}.\", \n  .         list(index = 1L, size = 5L, expected_size = 30146L), \n  .         \"dplyr:::filter_incompatible_size\"), rlang::exec(abort, \n  .         message = message, class = class, !!!.envir)), parents = c(0L, \n  . 1L, 2L, 3L, 4L, 5L, 6L, 7L, 6L, 9L, 10L, 4L, 12L, 13L, 13L, 15L, \n  . 16L, 17L, 18L, 19L, 13L, 13L, 13L, 23L, 0L, 25L, 25L, 27L, 28L, \n  . 29L, 30L, 30L, 32L, 32L, 34L, 35L, 35L, 0L, 38L), ids = c(\"0x7fc6254fee08\", \n  . \"0x7fc627046ba0\", \"0x7fc62f17d2c8\", \"0x7fc630839c48\", \"0x7fc62deb8b00\", \n  . \"0x7fc62deafd38\", \"0x7fc62deb05c0\", \"0x7fc62deb0a20\", \"0x7fc62deadd70\", \n  . \"0x7fc62deae2b0\", \"0x7fc62deae780\", \"0x7fc62deaf120\", \"0x7fc62ddd7dc0\", \n  . \"0x7fc62ddc30c8\", \"0x7fc62ddc31e0\", \"0x7fc62ddc3560\", \"0x7fc62ddc3870\", \n  . \"0x7fc62ddc03a8\", \"0x7fc62ddc08b0\", \"0x7fc62ddc0d80\", \"0x7fc62ddc1598\", \n  . \"0x7fc62ddbded8\", \"0x7fc62ddbe488\", \"global\", \"0x7fc62ddbe878\", \n  . \"0x7fc62ddb20d0\", \"0x7fc62ddb2450\", \"0x7fc62ddbed80\", \"0x7fc62ddb26b8\", \n  . \"0x7fc62ddb2990\", \"0x7fc62ddb2b50\", \"0x7fc62ddb2ca0\", \"0x7fc62ddb2fb0\", \n  . \"0x7fc62ddafa40\", \"0x7fc62ddafca8\", \"0x7fc62f3833b8\", \"0x7fc62f37ec48\", \n  . \"0x7fc62f377a48\", \"0x7fc62dcd7560\"), indices = 1:39), class = \"rlang_trace\"), \n  .     parent = NULL, index = 1L, size = 5L, expected_size = 30146L), class = c(\"dplyr:::filter_incompatible_size\", \n  . \"rlang_error\", \"error\", \"condition\")))",
      "20. abort(c(cnd_bullet_header(), x = conditionMessage(e), i = cnd_bullet_input_info(), \n  .     i = cnd_bullet_cur_group_label()), class = \"dplyr_error\")",
      "21. signal_abort(cnd)"
     ]
    }
   ],
   "source": [
    "countdata = countdata%>%\n",
    "    filter(colnames(countdata)[1:5] %in% c('MH112', 'MH113', 'MH114', 'MH115', 'MH116', 'MH117', 'MH118','MH101',\n",
    "                                           'MH102', 'MH103', 'MH104', 'MH105','MH106'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'X1'</li><li>'M1_RedOakSmolder'</li><li>'M10_PeatSmolder'</li><li>'M101_Saline'</li><li>'M107_RedOakFlame'</li><li>'M11_PeatSmolder'</li><li>'M113_PeatFlame'</li><li>'M119_PineNeedlesFlame'</li><li>'M12_PeatSmolder'</li><li>'M125_PineFlame'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'X1'\n",
       "\\item 'M1\\_RedOakSmolder'\n",
       "\\item 'M10\\_PeatSmolder'\n",
       "\\item 'M101\\_Saline'\n",
       "\\item 'M107\\_RedOakFlame'\n",
       "\\item 'M11\\_PeatSmolder'\n",
       "\\item 'M113\\_PeatFlame'\n",
       "\\item 'M119\\_PineNeedlesFlame'\n",
       "\\item 'M12\\_PeatSmolder'\n",
       "\\item 'M125\\_PineFlame'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'X1'\n",
       "2. 'M1_RedOakSmolder'\n",
       "3. 'M10_PeatSmolder'\n",
       "4. 'M101_Saline'\n",
       "5. 'M107_RedOakFlame'\n",
       "6. 'M11_PeatSmolder'\n",
       "7. 'M113_PeatFlame'\n",
       "8. 'M119_PineNeedlesFlame'\n",
       "9. 'M12_PeatSmolder'\n",
       "10. 'M125_PineFlame'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"X1\"                    \"M1_RedOakSmolder\"      \"M10_PeatSmolder\"      \n",
       " [4] \"M101_Saline\"           \"M107_RedOakFlame\"      \"M11_PeatSmolder\"      \n",
       " [7] \"M113_PeatFlame\"        \"M119_PineNeedlesFlame\" \"M12_PeatSmolder\"      \n",
       "[10] \"M125_PineFlame\"       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(countdata[1:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'X1'</li><li>'M1_RedOakSmolder'</li><li>'M10_PeatSmolder'</li><li>'M101_Saline'</li><li>'M107_RedOakFlame'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'X1'\n",
       "\\item 'M1\\_RedOakSmolder'\n",
       "\\item 'M10\\_PeatSmolder'\n",
       "\\item 'M101\\_Saline'\n",
       "\\item 'M107\\_RedOakFlame'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'X1'\n",
       "2. 'M1_RedOakSmolder'\n",
       "3. 'M10_PeatSmolder'\n",
       "4. 'M101_Saline'\n",
       "5. 'M107_RedOakFlame'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"X1\"               \"M1_RedOakSmolder\" \"M10_PeatSmolder\"  \"M101_Saline\"     \n",
       "[5] \"M107_RedOakFlame\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(countdata[1:10])[1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in substr(colname): argument \"start\" is missing, with no default\n",
     "output_type": "error",
     "traceback": [
      "Error in substr(colname): argument \"start\" is missing, with no default\nTraceback:\n",
      "1. substr(colname)"
     ]
    }
   ],
   "source": [
    "#only interested in comparing mouse heart peat flame (MH112-MH118) vs. mouse heart saline samples (MH101-MH106)\n",
    "for (i in 1:length(colnames(countdata))){\n",
    "#     if(i < 10){\n",
    "#         print(i)\n",
    "#         print(colnames(countdata)[i])\n",
    "#     }\n",
    "    colname = colnames(countdata)[i]\n",
    "    for (j in 1:length(colname)){\n",
    "        if (substr(colname,j,j+1) == 'MH'){ #saline\n",
    "            print(colname)\n",
    "            print(j)\n",
    "            #mouse heart peat flame\n",
    "#             if (substr(as.numeric(colname), j+3,j+6) <= 118){ #| substr(as.numeric(colname), j,j+3) >= 112){ \n",
    "#                 print(colname)\n",
    "#             }\n",
    "            \n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'M1_R'"
      ],
      "text/latex": [
       "'M1\\_R'"
      ],
      "text/markdown": [
       "'M1_R'"
      ],
      "text/plain": [
       "[1] \"M1_R\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a = colnames(countdata[2])\n",
    "substr(a,1,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'her'"
      ],
      "text/latex": [
       "'her'"
      ],
      "text/markdown": [
       "'her'"
      ],
      "text/plain": [
       "[1] \"her\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "string = 'hershey'\n",
    "substr(string,1,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in strsplit(string): argument \"split\" is missing, with no default\n",
     "output_type": "error",
     "traceback": [
      "Error in strsplit(string): argument \"split\" is missing, with no default\nTraceback:\n",
      "1. strsplit(string)"
     ]
    }
   ],
   "source": [
    "for (i in 1:length(strsplit(string))){\n",
    "    print(i)\n",
    "    print(string[i])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Missing column names filled in: 'X1' [1]”\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\n",
      "See spec(...) for full column specifications.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>30146</li><li>171</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 30146\n",
       "\\item 171\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 30146\n",
       "2. 171\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 30146   171"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 3 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>X1</th><th scope=col>M1_RedOakSmolder</th><th scope=col>M10_PeatSmolder</th><th scope=col>M101_Saline</th><th scope=col>M107_RedOakFlame</th><th scope=col>M11_PeatSmolder</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Bcat1_29039 </td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><td>Laptm5_29040</td><td>0</td><td>2</td><td>2</td><td>1</td><td>4</td></tr>\n",
       "\t<tr><td>Skp2_29041  </td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 3 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " X1 & M1\\_RedOakSmolder & M10\\_PeatSmolder & M101\\_Saline & M107\\_RedOakFlame & M11\\_PeatSmolder\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Bcat1\\_29039  & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\t Laptm5\\_29040 & 0 & 2 & 2 & 1 & 4\\\\\n",
       "\t Skp2\\_29041   & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 3 × 6\n",
       "\n",
       "| X1 &lt;chr&gt; | M1_RedOakSmolder &lt;dbl&gt; | M10_PeatSmolder &lt;dbl&gt; | M101_Saline &lt;dbl&gt; | M107_RedOakFlame &lt;dbl&gt; | M11_PeatSmolder &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| Bcat1_29039  | 0 | 0 | 0 | 0 | 0 |\n",
       "| Laptm5_29040 | 0 | 2 | 2 | 1 | 4 |\n",
       "| Skp2_29041   | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  X1           M1_RedOakSmolder M10_PeatSmolder M101_Saline M107_RedOakFlame\n",
       "1 Bcat1_29039  0                0               0           0               \n",
       "2 Laptm5_29040 0                2               2           1               \n",
       "3 Skp2_29041   0                0               0           0               \n",
       "  M11_PeatSmolder\n",
       "1 0              \n",
       "2 4              \n",
       "3 0              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   Mode   FALSE \n",
       "logical   30146 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(countdata) #171 samples (plus Gene ID column), 30146 genes\n",
    "# visualize this data quickly by viewing top left corner:\n",
    "countdata[1:3,1:6]\n",
    "\n",
    "# Check for duplicate gene IDs in the countdata \n",
    "Dups <- duplicated(countdata[,1])\n",
    "summary(Dups)\n",
    "# Not an issue for this wildfire dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Gene'</li><li>'M100_EucalyptusSmolder'</li><li>'M101_Saline'</li><li>'M102_Saline'</li><li>'M103_Saline'</li><li>'M104_Saline'</li><li>'M105_Saline'</li><li>'M106_Saline'</li><li>'M107_RedOakFlame'</li><li>'M108_RedOakFlame'</li><li>'M109_RedOakFlame'</li><li>'M10_PeatSmolder'</li><li>'M110_RedOakFlame'</li><li>'M111_RedOakFlame'</li><li>'M112_RedOakFlame'</li><li>'M113_PeatFlame'</li><li>'M114_PeatFlame'</li><li>'M115_PeatFlame'</li><li>'M116_PeatFlame'</li><li>'M117_PeatFlame'</li><li>'M118_PeatFlame'</li><li>'M119_PineNeedlesFlame'</li><li>'M11_PeatSmolder'</li><li>'M120_PineNeedlesFlame'</li><li>'M121_PineNeedlesFlame'</li><li>'M122_PineNeedlesFlame'</li><li>'M123_PineNeedlesFlame'</li><li>'M124_PineNeedlesFlame'</li><li>'M125_PineFlame'</li><li>'M126_PineFlame'</li><li>'M127_PineFlame'</li><li>'M128_PineFlame'</li><li>'M129_PineFlame'</li><li>'M12_PeatSmolder'</li><li>'M130_PineFlame'</li><li>'M131_PineFlame'</li><li>'M132_PineFlame'</li><li>'M133_PineFlame'</li><li>'M134_PineFlame'</li><li>'M135_PineFlame'</li><li>'M136_PineFlame'</li><li>'M137_LPS'</li><li>'M138_LPS'</li><li>'M139_LPS'</li><li>'M13_PeatSmolder'</li><li>'M140_LPS'</li><li>'M14_PeatSmolder'</li><li>'M15_PeatSmolder'</li><li>'M16_PeatSmolder'</li><li>'M17_PeatSmolder'</li><li>'M18_PeatSmolder'</li><li>'M19_PineSmolder'</li><li>'M1_RedOakSmolder'</li><li>'M20_PineSmolder'</li><li>'M21_PineSmolder'</li><li>'M22_PineSmolder'</li><li>'M23_PineSmolder'</li><li>'M24_PineSmolder'</li><li>'M25_EucalyptusSmolder'</li><li>'M26_EucalyptusSmolder'</li><li>'M27_EucalyptusSmolder'</li><li>'M28_EucalyptusSmolder'</li><li>'M29_EucalyptusSmolder'</li><li>'M2_RedOakSmolder'</li><li>'M30_EucalyptusSmolder'</li><li>'M31_Saline'</li><li>'M32_Saline'</li><li>'M33_Saline'</li><li>'M34_Saline'</li><li>'M35_Saline'</li><li>'M36_Saline'</li><li>'M37_RedOakFlame'</li><li>'M38_RedOakFlame'</li><li>'M39_RedOakFlame'</li><li>'M3_RedOakSmolder'</li><li>'M40_RedOakFlame'</li><li>'M41_RedOakFlame'</li><li>'M42_RedOakFlame'</li><li>'M43_PeatFlame'</li><li>'M44_PeatFlame'</li><li>'M45_PeatFlame'</li><li>'M46_PeatFlame'</li><li>'M47_PeatFlame'</li><li>'M48_PeatFlame'</li><li>'M49_PineNeedlesFlame'</li><li>'M4_RedOakSmolder'</li><li>'M50_PineNeedlesFlame'</li><li>'M51_PineNeedlesFlame'</li><li>'M52_PineNeedlesFlame'</li><li>'M53_PineNeedlesFlame'</li><li>'M54_PineNeedlesFlame'</li><li>'M55_PineFlame'</li><li>'M56_PineFlame'</li><li>'M57_PineFlame'</li><li>'M58_PineFlame'</li><li>'M59_PineFlame'</li><li>'M5_RedOakSmolder'</li><li>'M60_PineFlame'</li><li>'M61_EucalyptusFlame'</li><li>'M62_EucalyptusFlame'</li><li>'M63_EucalyptusFlame'</li><li>'M64_EucalyptusFlame'</li><li>'M65_EucalyptusFlame'</li><li>'M66_EucalyptusFlame'</li><li>'M67_LPS'</li><li>'M68_LPS'</li><li>'M69_LPS'</li><li>'M6_RedOakSmolder'</li><li>'M70_LPS'</li><li>'M71_RedOakSmolder'</li><li>'M72_RedOakSmolder'</li><li>'M73_RedOakSmolder'</li><li>'M74_RedOakSmolder'</li><li>'M75_RedOakSmolder'</li><li>'M76_RedOakSmolder'</li><li>'M77_PeatSmolder'</li><li>'M78_PeatSmolder'</li><li>'M79_PeatSmolder'</li><li>'M7_PeatSmolder'</li><li>'M80_PeatSmolder'</li><li>'M81_PeatSmolder'</li><li>'M82_PeatSmolder'</li><li>'M83_PineNeedlesSmolder'</li><li>'M84_PineNeedlesSmolder'</li><li>'M85_PineNeedlesSmolder'</li><li>'M86_PineNeedlesSmolder'</li><li>'M87_PineNeedlesSmolder'</li><li>'M88_PineNeedlesSmolder'</li><li>'M89_PineSmolder'</li><li>'M8_PeatSmolder'</li><li>'M90_PineSmolder'</li><li>'M91_PineSmolder'</li><li>'M92_PineSmolder'</li><li>'M93_PineSmolder'</li><li>'M94_PineSmolder'</li><li>'M95_EucalyptusSmolder'</li><li>'M96_EucalyptusSmolder'</li><li>'M97_EucalyptusSmolder'</li><li>'M98_EucalyptusSmolder'</li><li>'M99_EucalyptusSmolder'</li><li>'M9_PeatSmolder'</li><li>'MH101_Saline'</li><li>'MH102_Saline'</li><li>'MH103_Saline'</li><li>'MH104_Saline'</li><li>'MH105_Saline'</li><li>'MH106_Saline'</li><li>'MH107_RedOakFlame'</li><li>'MH108_RedOakFlame'</li><li>'MH109_RedOakFlame'</li><li>'MH110_RedOakFlame'</li><li>'MH111_RedOakFlame'</li><li>'MH112_RedOakFlame'</li><li>'MH113_PeatFlame'</li><li>'MH114_PeatFlame'</li><li>'MH115_PeatFlame'</li><li>'MH116_PeatFlame'</li><li>'MH117_PeatFlame'</li><li>'MH118_PeatFlame'</li><li>'MH71_RedOakSmolder'</li><li>'MH72_RedOakSmolder'</li><li>'MH73_RedOakSmolder'</li><li>'MH74_RedOakSmolder'</li><li>'MH75_RedOakSmolder'</li><li>'MH76_RedOakSmolder'</li><li>'MH77_PeatSmolder'</li><li>'MH78_PeatSmolder'</li><li>'MH79_PeatSmolder'</li><li>'MH80_PeatSmolder'</li><li>'MH81_PeatSmolder'</li><li>'MH82_PeatSmolder'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Gene'\n",
       "\\item 'M100\\_EucalyptusSmolder'\n",
       "\\item 'M101\\_Saline'\n",
       "\\item 'M102\\_Saline'\n",
       "\\item 'M103\\_Saline'\n",
       "\\item 'M104\\_Saline'\n",
       "\\item 'M105\\_Saline'\n",
       "\\item 'M106\\_Saline'\n",
       "\\item 'M107\\_RedOakFlame'\n",
       "\\item 'M108\\_RedOakFlame'\n",
       "\\item 'M109\\_RedOakFlame'\n",
       "\\item 'M10\\_PeatSmolder'\n",
       "\\item 'M110\\_RedOakFlame'\n",
       "\\item 'M111\\_RedOakFlame'\n",
       "\\item 'M112\\_RedOakFlame'\n",
       "\\item 'M113\\_PeatFlame'\n",
       "\\item 'M114\\_PeatFlame'\n",
       "\\item 'M115\\_PeatFlame'\n",
       "\\item 'M116\\_PeatFlame'\n",
       "\\item 'M117\\_PeatFlame'\n",
       "\\item 'M118\\_PeatFlame'\n",
       "\\item 'M119\\_PineNeedlesFlame'\n",
       "\\item 'M11\\_PeatSmolder'\n",
       "\\item 'M120\\_PineNeedlesFlame'\n",
       "\\item 'M121\\_PineNeedlesFlame'\n",
       "\\item 'M122\\_PineNeedlesFlame'\n",
       "\\item 'M123\\_PineNeedlesFlame'\n",
       "\\item 'M124\\_PineNeedlesFlame'\n",
       "\\item 'M125\\_PineFlame'\n",
       "\\item 'M126\\_PineFlame'\n",
       "\\item 'M127\\_PineFlame'\n",
       "\\item 'M128\\_PineFlame'\n",
       "\\item 'M129\\_PineFlame'\n",
       "\\item 'M12\\_PeatSmolder'\n",
       "\\item 'M130\\_PineFlame'\n",
       "\\item 'M131\\_PineFlame'\n",
       "\\item 'M132\\_PineFlame'\n",
       "\\item 'M133\\_PineFlame'\n",
       "\\item 'M134\\_PineFlame'\n",
       "\\item 'M135\\_PineFlame'\n",
       "\\item 'M136\\_PineFlame'\n",
       "\\item 'M137\\_LPS'\n",
       "\\item 'M138\\_LPS'\n",
       "\\item 'M139\\_LPS'\n",
       "\\item 'M13\\_PeatSmolder'\n",
       "\\item 'M140\\_LPS'\n",
       "\\item 'M14\\_PeatSmolder'\n",
       "\\item 'M15\\_PeatSmolder'\n",
       "\\item 'M16\\_PeatSmolder'\n",
       "\\item 'M17\\_PeatSmolder'\n",
       "\\item 'M18\\_PeatSmolder'\n",
       "\\item 'M19\\_PineSmolder'\n",
       "\\item 'M1\\_RedOakSmolder'\n",
       "\\item 'M20\\_PineSmolder'\n",
       "\\item 'M21\\_PineSmolder'\n",
       "\\item 'M22\\_PineSmolder'\n",
       "\\item 'M23\\_PineSmolder'\n",
       "\\item 'M24\\_PineSmolder'\n",
       "\\item 'M25\\_EucalyptusSmolder'\n",
       "\\item 'M26\\_EucalyptusSmolder'\n",
       "\\item 'M27\\_EucalyptusSmolder'\n",
       "\\item 'M28\\_EucalyptusSmolder'\n",
       "\\item 'M29\\_EucalyptusSmolder'\n",
       "\\item 'M2\\_RedOakSmolder'\n",
       "\\item 'M30\\_EucalyptusSmolder'\n",
       "\\item 'M31\\_Saline'\n",
       "\\item 'M32\\_Saline'\n",
       "\\item 'M33\\_Saline'\n",
       "\\item 'M34\\_Saline'\n",
       "\\item 'M35\\_Saline'\n",
       "\\item 'M36\\_Saline'\n",
       "\\item 'M37\\_RedOakFlame'\n",
       "\\item 'M38\\_RedOakFlame'\n",
       "\\item 'M39\\_RedOakFlame'\n",
       "\\item 'M3\\_RedOakSmolder'\n",
       "\\item 'M40\\_RedOakFlame'\n",
       "\\item 'M41\\_RedOakFlame'\n",
       "\\item 'M42\\_RedOakFlame'\n",
       "\\item 'M43\\_PeatFlame'\n",
       "\\item 'M44\\_PeatFlame'\n",
       "\\item 'M45\\_PeatFlame'\n",
       "\\item 'M46\\_PeatFlame'\n",
       "\\item 'M47\\_PeatFlame'\n",
       "\\item 'M48\\_PeatFlame'\n",
       "\\item 'M49\\_PineNeedlesFlame'\n",
       "\\item 'M4\\_RedOakSmolder'\n",
       "\\item 'M50\\_PineNeedlesFlame'\n",
       "\\item 'M51\\_PineNeedlesFlame'\n",
       "\\item 'M52\\_PineNeedlesFlame'\n",
       "\\item 'M53\\_PineNeedlesFlame'\n",
       "\\item 'M54\\_PineNeedlesFlame'\n",
       "\\item 'M55\\_PineFlame'\n",
       "\\item 'M56\\_PineFlame'\n",
       "\\item 'M57\\_PineFlame'\n",
       "\\item 'M58\\_PineFlame'\n",
       "\\item 'M59\\_PineFlame'\n",
       "\\item 'M5\\_RedOakSmolder'\n",
       "\\item 'M60\\_PineFlame'\n",
       "\\item 'M61\\_EucalyptusFlame'\n",
       "\\item 'M62\\_EucalyptusFlame'\n",
       "\\item 'M63\\_EucalyptusFlame'\n",
       "\\item 'M64\\_EucalyptusFlame'\n",
       "\\item 'M65\\_EucalyptusFlame'\n",
       "\\item 'M66\\_EucalyptusFlame'\n",
       "\\item 'M67\\_LPS'\n",
       "\\item 'M68\\_LPS'\n",
       "\\item 'M69\\_LPS'\n",
       "\\item 'M6\\_RedOakSmolder'\n",
       "\\item 'M70\\_LPS'\n",
       "\\item 'M71\\_RedOakSmolder'\n",
       "\\item 'M72\\_RedOakSmolder'\n",
       "\\item 'M73\\_RedOakSmolder'\n",
       "\\item 'M74\\_RedOakSmolder'\n",
       "\\item 'M75\\_RedOakSmolder'\n",
       "\\item 'M76\\_RedOakSmolder'\n",
       "\\item 'M77\\_PeatSmolder'\n",
       "\\item 'M78\\_PeatSmolder'\n",
       "\\item 'M79\\_PeatSmolder'\n",
       "\\item 'M7\\_PeatSmolder'\n",
       "\\item 'M80\\_PeatSmolder'\n",
       "\\item 'M81\\_PeatSmolder'\n",
       "\\item 'M82\\_PeatSmolder'\n",
       "\\item 'M83\\_PineNeedlesSmolder'\n",
       "\\item 'M84\\_PineNeedlesSmolder'\n",
       "\\item 'M85\\_PineNeedlesSmolder'\n",
       "\\item 'M86\\_PineNeedlesSmolder'\n",
       "\\item 'M87\\_PineNeedlesSmolder'\n",
       "\\item 'M88\\_PineNeedlesSmolder'\n",
       "\\item 'M89\\_PineSmolder'\n",
       "\\item 'M8\\_PeatSmolder'\n",
       "\\item 'M90\\_PineSmolder'\n",
       "\\item 'M91\\_PineSmolder'\n",
       "\\item 'M92\\_PineSmolder'\n",
       "\\item 'M93\\_PineSmolder'\n",
       "\\item 'M94\\_PineSmolder'\n",
       "\\item 'M95\\_EucalyptusSmolder'\n",
       "\\item 'M96\\_EucalyptusSmolder'\n",
       "\\item 'M97\\_EucalyptusSmolder'\n",
       "\\item 'M98\\_EucalyptusSmolder'\n",
       "\\item 'M99\\_EucalyptusSmolder'\n",
       "\\item 'M9\\_PeatSmolder'\n",
       "\\item 'MH101\\_Saline'\n",
       "\\item 'MH102\\_Saline'\n",
       "\\item 'MH103\\_Saline'\n",
       "\\item 'MH104\\_Saline'\n",
       "\\item 'MH105\\_Saline'\n",
       "\\item 'MH106\\_Saline'\n",
       "\\item 'MH107\\_RedOakFlame'\n",
       "\\item 'MH108\\_RedOakFlame'\n",
       "\\item 'MH109\\_RedOakFlame'\n",
       "\\item 'MH110\\_RedOakFlame'\n",
       "\\item 'MH111\\_RedOakFlame'\n",
       "\\item 'MH112\\_RedOakFlame'\n",
       "\\item 'MH113\\_PeatFlame'\n",
       "\\item 'MH114\\_PeatFlame'\n",
       "\\item 'MH115\\_PeatFlame'\n",
       "\\item 'MH116\\_PeatFlame'\n",
       "\\item 'MH117\\_PeatFlame'\n",
       "\\item 'MH118\\_PeatFlame'\n",
       "\\item 'MH71\\_RedOakSmolder'\n",
       "\\item 'MH72\\_RedOakSmolder'\n",
       "\\item 'MH73\\_RedOakSmolder'\n",
       "\\item 'MH74\\_RedOakSmolder'\n",
       "\\item 'MH75\\_RedOakSmolder'\n",
       "\\item 'MH76\\_RedOakSmolder'\n",
       "\\item 'MH77\\_PeatSmolder'\n",
       "\\item 'MH78\\_PeatSmolder'\n",
       "\\item 'MH79\\_PeatSmolder'\n",
       "\\item 'MH80\\_PeatSmolder'\n",
       "\\item 'MH81\\_PeatSmolder'\n",
       "\\item 'MH82\\_PeatSmolder'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Gene'\n",
       "2. 'M100_EucalyptusSmolder'\n",
       "3. 'M101_Saline'\n",
       "4. 'M102_Saline'\n",
       "5. 'M103_Saline'\n",
       "6. 'M104_Saline'\n",
       "7. 'M105_Saline'\n",
       "8. 'M106_Saline'\n",
       "9. 'M107_RedOakFlame'\n",
       "10. 'M108_RedOakFlame'\n",
       "11. 'M109_RedOakFlame'\n",
       "12. 'M10_PeatSmolder'\n",
       "13. 'M110_RedOakFlame'\n",
       "14. 'M111_RedOakFlame'\n",
       "15. 'M112_RedOakFlame'\n",
       "16. 'M113_PeatFlame'\n",
       "17. 'M114_PeatFlame'\n",
       "18. 'M115_PeatFlame'\n",
       "19. 'M116_PeatFlame'\n",
       "20. 'M117_PeatFlame'\n",
       "21. 'M118_PeatFlame'\n",
       "22. 'M119_PineNeedlesFlame'\n",
       "23. 'M11_PeatSmolder'\n",
       "24. 'M120_PineNeedlesFlame'\n",
       "25. 'M121_PineNeedlesFlame'\n",
       "26. 'M122_PineNeedlesFlame'\n",
       "27. 'M123_PineNeedlesFlame'\n",
       "28. 'M124_PineNeedlesFlame'\n",
       "29. 'M125_PineFlame'\n",
       "30. 'M126_PineFlame'\n",
       "31. 'M127_PineFlame'\n",
       "32. 'M128_PineFlame'\n",
       "33. 'M129_PineFlame'\n",
       "34. 'M12_PeatSmolder'\n",
       "35. 'M130_PineFlame'\n",
       "36. 'M131_PineFlame'\n",
       "37. 'M132_PineFlame'\n",
       "38. 'M133_PineFlame'\n",
       "39. 'M134_PineFlame'\n",
       "40. 'M135_PineFlame'\n",
       "41. 'M136_PineFlame'\n",
       "42. 'M137_LPS'\n",
       "43. 'M138_LPS'\n",
       "44. 'M139_LPS'\n",
       "45. 'M13_PeatSmolder'\n",
       "46. 'M140_LPS'\n",
       "47. 'M14_PeatSmolder'\n",
       "48. 'M15_PeatSmolder'\n",
       "49. 'M16_PeatSmolder'\n",
       "50. 'M17_PeatSmolder'\n",
       "51. 'M18_PeatSmolder'\n",
       "52. 'M19_PineSmolder'\n",
       "53. 'M1_RedOakSmolder'\n",
       "54. 'M20_PineSmolder'\n",
       "55. 'M21_PineSmolder'\n",
       "56. 'M22_PineSmolder'\n",
       "57. 'M23_PineSmolder'\n",
       "58. 'M24_PineSmolder'\n",
       "59. 'M25_EucalyptusSmolder'\n",
       "60. 'M26_EucalyptusSmolder'\n",
       "61. 'M27_EucalyptusSmolder'\n",
       "62. 'M28_EucalyptusSmolder'\n",
       "63. 'M29_EucalyptusSmolder'\n",
       "64. 'M2_RedOakSmolder'\n",
       "65. 'M30_EucalyptusSmolder'\n",
       "66. 'M31_Saline'\n",
       "67. 'M32_Saline'\n",
       "68. 'M33_Saline'\n",
       "69. 'M34_Saline'\n",
       "70. 'M35_Saline'\n",
       "71. 'M36_Saline'\n",
       "72. 'M37_RedOakFlame'\n",
       "73. 'M38_RedOakFlame'\n",
       "74. 'M39_RedOakFlame'\n",
       "75. 'M3_RedOakSmolder'\n",
       "76. 'M40_RedOakFlame'\n",
       "77. 'M41_RedOakFlame'\n",
       "78. 'M42_RedOakFlame'\n",
       "79. 'M43_PeatFlame'\n",
       "80. 'M44_PeatFlame'\n",
       "81. 'M45_PeatFlame'\n",
       "82. 'M46_PeatFlame'\n",
       "83. 'M47_PeatFlame'\n",
       "84. 'M48_PeatFlame'\n",
       "85. 'M49_PineNeedlesFlame'\n",
       "86. 'M4_RedOakSmolder'\n",
       "87. 'M50_PineNeedlesFlame'\n",
       "88. 'M51_PineNeedlesFlame'\n",
       "89. 'M52_PineNeedlesFlame'\n",
       "90. 'M53_PineNeedlesFlame'\n",
       "91. 'M54_PineNeedlesFlame'\n",
       "92. 'M55_PineFlame'\n",
       "93. 'M56_PineFlame'\n",
       "94. 'M57_PineFlame'\n",
       "95. 'M58_PineFlame'\n",
       "96. 'M59_PineFlame'\n",
       "97. 'M5_RedOakSmolder'\n",
       "98. 'M60_PineFlame'\n",
       "99. 'M61_EucalyptusFlame'\n",
       "100. 'M62_EucalyptusFlame'\n",
       "101. 'M63_EucalyptusFlame'\n",
       "102. 'M64_EucalyptusFlame'\n",
       "103. 'M65_EucalyptusFlame'\n",
       "104. 'M66_EucalyptusFlame'\n",
       "105. 'M67_LPS'\n",
       "106. 'M68_LPS'\n",
       "107. 'M69_LPS'\n",
       "108. 'M6_RedOakSmolder'\n",
       "109. 'M70_LPS'\n",
       "110. 'M71_RedOakSmolder'\n",
       "111. 'M72_RedOakSmolder'\n",
       "112. 'M73_RedOakSmolder'\n",
       "113. 'M74_RedOakSmolder'\n",
       "114. 'M75_RedOakSmolder'\n",
       "115. 'M76_RedOakSmolder'\n",
       "116. 'M77_PeatSmolder'\n",
       "117. 'M78_PeatSmolder'\n",
       "118. 'M79_PeatSmolder'\n",
       "119. 'M7_PeatSmolder'\n",
       "120. 'M80_PeatSmolder'\n",
       "121. 'M81_PeatSmolder'\n",
       "122. 'M82_PeatSmolder'\n",
       "123. 'M83_PineNeedlesSmolder'\n",
       "124. 'M84_PineNeedlesSmolder'\n",
       "125. 'M85_PineNeedlesSmolder'\n",
       "126. 'M86_PineNeedlesSmolder'\n",
       "127. 'M87_PineNeedlesSmolder'\n",
       "128. 'M88_PineNeedlesSmolder'\n",
       "129. 'M89_PineSmolder'\n",
       "130. 'M8_PeatSmolder'\n",
       "131. 'M90_PineSmolder'\n",
       "132. 'M91_PineSmolder'\n",
       "133. 'M92_PineSmolder'\n",
       "134. 'M93_PineSmolder'\n",
       "135. 'M94_PineSmolder'\n",
       "136. 'M95_EucalyptusSmolder'\n",
       "137. 'M96_EucalyptusSmolder'\n",
       "138. 'M97_EucalyptusSmolder'\n",
       "139. 'M98_EucalyptusSmolder'\n",
       "140. 'M99_EucalyptusSmolder'\n",
       "141. 'M9_PeatSmolder'\n",
       "142. 'MH101_Saline'\n",
       "143. 'MH102_Saline'\n",
       "144. 'MH103_Saline'\n",
       "145. 'MH104_Saline'\n",
       "146. 'MH105_Saline'\n",
       "147. 'MH106_Saline'\n",
       "148. 'MH107_RedOakFlame'\n",
       "149. 'MH108_RedOakFlame'\n",
       "150. 'MH109_RedOakFlame'\n",
       "151. 'MH110_RedOakFlame'\n",
       "152. 'MH111_RedOakFlame'\n",
       "153. 'MH112_RedOakFlame'\n",
       "154. 'MH113_PeatFlame'\n",
       "155. 'MH114_PeatFlame'\n",
       "156. 'MH115_PeatFlame'\n",
       "157. 'MH116_PeatFlame'\n",
       "158. 'MH117_PeatFlame'\n",
       "159. 'MH118_PeatFlame'\n",
       "160. 'MH71_RedOakSmolder'\n",
       "161. 'MH72_RedOakSmolder'\n",
       "162. 'MH73_RedOakSmolder'\n",
       "163. 'MH74_RedOakSmolder'\n",
       "164. 'MH75_RedOakSmolder'\n",
       "165. 'MH76_RedOakSmolder'\n",
       "166. 'MH77_PeatSmolder'\n",
       "167. 'MH78_PeatSmolder'\n",
       "168. 'MH79_PeatSmolder'\n",
       "169. 'MH80_PeatSmolder'\n",
       "170. 'MH81_PeatSmolder'\n",
       "171. 'MH82_PeatSmolder'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"Gene\"                   \"M100_EucalyptusSmolder\" \"M101_Saline\"           \n",
       "  [4] \"M102_Saline\"            \"M103_Saline\"            \"M104_Saline\"           \n",
       "  [7] \"M105_Saline\"            \"M106_Saline\"            \"M107_RedOakFlame\"      \n",
       " [10] \"M108_RedOakFlame\"       \"M109_RedOakFlame\"       \"M10_PeatSmolder\"       \n",
       " [13] \"M110_RedOakFlame\"       \"M111_RedOakFlame\"       \"M112_RedOakFlame\"      \n",
       " [16] \"M113_PeatFlame\"         \"M114_PeatFlame\"         \"M115_PeatFlame\"        \n",
       " [19] \"M116_PeatFlame\"         \"M117_PeatFlame\"         \"M118_PeatFlame\"        \n",
       " [22] \"M119_PineNeedlesFlame\"  \"M11_PeatSmolder\"        \"M120_PineNeedlesFlame\" \n",
       " [25] \"M121_PineNeedlesFlame\"  \"M122_PineNeedlesFlame\"  \"M123_PineNeedlesFlame\" \n",
       " [28] \"M124_PineNeedlesFlame\"  \"M125_PineFlame\"         \"M126_PineFlame\"        \n",
       " [31] \"M127_PineFlame\"         \"M128_PineFlame\"         \"M129_PineFlame\"        \n",
       " [34] \"M12_PeatSmolder\"        \"M130_PineFlame\"         \"M131_PineFlame\"        \n",
       " [37] \"M132_PineFlame\"         \"M133_PineFlame\"         \"M134_PineFlame\"        \n",
       " [40] \"M135_PineFlame\"         \"M136_PineFlame\"         \"M137_LPS\"              \n",
       " [43] \"M138_LPS\"               \"M139_LPS\"               \"M13_PeatSmolder\"       \n",
       " [46] \"M140_LPS\"               \"M14_PeatSmolder\"        \"M15_PeatSmolder\"       \n",
       " [49] \"M16_PeatSmolder\"        \"M17_PeatSmolder\"        \"M18_PeatSmolder\"       \n",
       " [52] \"M19_PineSmolder\"        \"M1_RedOakSmolder\"       \"M20_PineSmolder\"       \n",
       " [55] \"M21_PineSmolder\"        \"M22_PineSmolder\"        \"M23_PineSmolder\"       \n",
       " [58] \"M24_PineSmolder\"        \"M25_EucalyptusSmolder\"  \"M26_EucalyptusSmolder\" \n",
       " [61] \"M27_EucalyptusSmolder\"  \"M28_EucalyptusSmolder\"  \"M29_EucalyptusSmolder\" \n",
       " [64] \"M2_RedOakSmolder\"       \"M30_EucalyptusSmolder\"  \"M31_Saline\"            \n",
       " [67] \"M32_Saline\"             \"M33_Saline\"             \"M34_Saline\"            \n",
       " [70] \"M35_Saline\"             \"M36_Saline\"             \"M37_RedOakFlame\"       \n",
       " [73] \"M38_RedOakFlame\"        \"M39_RedOakFlame\"        \"M3_RedOakSmolder\"      \n",
       " [76] \"M40_RedOakFlame\"        \"M41_RedOakFlame\"        \"M42_RedOakFlame\"       \n",
       " [79] \"M43_PeatFlame\"          \"M44_PeatFlame\"          \"M45_PeatFlame\"         \n",
       " [82] \"M46_PeatFlame\"          \"M47_PeatFlame\"          \"M48_PeatFlame\"         \n",
       " [85] \"M49_PineNeedlesFlame\"   \"M4_RedOakSmolder\"       \"M50_PineNeedlesFlame\"  \n",
       " [88] \"M51_PineNeedlesFlame\"   \"M52_PineNeedlesFlame\"   \"M53_PineNeedlesFlame\"  \n",
       " [91] \"M54_PineNeedlesFlame\"   \"M55_PineFlame\"          \"M56_PineFlame\"         \n",
       " [94] \"M57_PineFlame\"          \"M58_PineFlame\"          \"M59_PineFlame\"         \n",
       " [97] \"M5_RedOakSmolder\"       \"M60_PineFlame\"          \"M61_EucalyptusFlame\"   \n",
       "[100] \"M62_EucalyptusFlame\"    \"M63_EucalyptusFlame\"    \"M64_EucalyptusFlame\"   \n",
       "[103] \"M65_EucalyptusFlame\"    \"M66_EucalyptusFlame\"    \"M67_LPS\"               \n",
       "[106] \"M68_LPS\"                \"M69_LPS\"                \"M6_RedOakSmolder\"      \n",
       "[109] \"M70_LPS\"                \"M71_RedOakSmolder\"      \"M72_RedOakSmolder\"     \n",
       "[112] \"M73_RedOakSmolder\"      \"M74_RedOakSmolder\"      \"M75_RedOakSmolder\"     \n",
       "[115] \"M76_RedOakSmolder\"      \"M77_PeatSmolder\"        \"M78_PeatSmolder\"       \n",
       "[118] \"M79_PeatSmolder\"        \"M7_PeatSmolder\"         \"M80_PeatSmolder\"       \n",
       "[121] \"M81_PeatSmolder\"        \"M82_PeatSmolder\"        \"M83_PineNeedlesSmolder\"\n",
       "[124] \"M84_PineNeedlesSmolder\" \"M85_PineNeedlesSmolder\" \"M86_PineNeedlesSmolder\"\n",
       "[127] \"M87_PineNeedlesSmolder\" \"M88_PineNeedlesSmolder\" \"M89_PineSmolder\"       \n",
       "[130] \"M8_PeatSmolder\"         \"M90_PineSmolder\"        \"M91_PineSmolder\"       \n",
       "[133] \"M92_PineSmolder\"        \"M93_PineSmolder\"        \"M94_PineSmolder\"       \n",
       "[136] \"M95_EucalyptusSmolder\"  \"M96_EucalyptusSmolder\"  \"M97_EucalyptusSmolder\" \n",
       "[139] \"M98_EucalyptusSmolder\"  \"M99_EucalyptusSmolder\"  \"M9_PeatSmolder\"        \n",
       "[142] \"MH101_Saline\"           \"MH102_Saline\"           \"MH103_Saline\"          \n",
       "[145] \"MH104_Saline\"           \"MH105_Saline\"           \"MH106_Saline\"          \n",
       "[148] \"MH107_RedOakFlame\"      \"MH108_RedOakFlame\"      \"MH109_RedOakFlame\"     \n",
       "[151] \"MH110_RedOakFlame\"      \"MH111_RedOakFlame\"      \"MH112_RedOakFlame\"     \n",
       "[154] \"MH113_PeatFlame\"        \"MH114_PeatFlame\"        \"MH115_PeatFlame\"       \n",
       "[157] \"MH116_PeatFlame\"        \"MH117_PeatFlame\"        \"MH118_PeatFlame\"       \n",
       "[160] \"MH71_RedOakSmolder\"     \"MH72_RedOakSmolder\"     \"MH73_RedOakSmolder\"    \n",
       "[163] \"MH74_RedOakSmolder\"     \"MH75_RedOakSmolder\"     \"MH76_RedOakSmolder\"    \n",
       "[166] \"MH77_PeatSmolder\"       \"MH78_PeatSmolder\"       \"MH79_PeatSmolder\"      \n",
       "[169] \"MH80_PeatSmolder\"       \"MH81_PeatSmolder\"       \"MH82_PeatSmolder\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(countdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>170</li><li>7</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 170\n",
       "\\item 7\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 170\n",
       "2. 7\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 170   7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A tibble: 3 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>SampleID_BioSpyderCountFile</th><th scope=col>PlateBatch</th><th scope=col>MouseID</th><th scope=col>MouseID_CAPS</th><th scope=col>FullTreatment</th><th scope=col>FullID</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Plate1-m1-RNA  </td><td>Plate1</td><td>m1  </td><td>M1  </td><td>RedOakSmolder</td><td>M1_RedOakSmolder</td></tr>\n",
       "\t<tr><td>Plate1-m10-RNA </td><td>Plate1</td><td>m10 </td><td>M10 </td><td>PeatSmolder  </td><td>M10_PeatSmolder </td></tr>\n",
       "\t<tr><td>Plate1-m101-RNA</td><td>Plate1</td><td>m101</td><td>M101</td><td>Saline       </td><td>M101_Saline     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 3 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " SampleID\\_BioSpyderCountFile & PlateBatch & MouseID & MouseID\\_CAPS & FullTreatment & FullID\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t Plate1-m1-RNA   & Plate1 & m1   & M1   & RedOakSmolder & M1\\_RedOakSmolder\\\\\n",
       "\t Plate1-m10-RNA  & Plate1 & m10  & M10  & PeatSmolder   & M10\\_PeatSmolder \\\\\n",
       "\t Plate1-m101-RNA & Plate1 & m101 & M101 & Saline        & M101\\_Saline     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 3 × 6\n",
       "\n",
       "| SampleID_BioSpyderCountFile &lt;chr&gt; | PlateBatch &lt;chr&gt; | MouseID &lt;chr&gt; | MouseID_CAPS &lt;chr&gt; | FullTreatment &lt;chr&gt; | FullID &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| Plate1-m1-RNA   | Plate1 | m1   | M1   | RedOakSmolder | M1_RedOakSmolder |\n",
       "| Plate1-m10-RNA  | Plate1 | m10  | M10  | PeatSmolder   | M10_PeatSmolder  |\n",
       "| Plate1-m101-RNA | Plate1 | m101 | M101 | Saline        | M101_Saline      |\n",
       "\n"
      ],
      "text/plain": [
       "  SampleID_BioSpyderCountFile PlateBatch MouseID MouseID_CAPS FullTreatment\n",
       "1 Plate1-m1-RNA               Plate1     m1      M1           RedOakSmolder\n",
       "2 Plate1-m10-RNA              Plate1     m10     M10          PeatSmolder  \n",
       "3 Plate1-m101-RNA             Plate1     m101    M101         Saline       \n",
       "  FullID          \n",
       "1 M1_RedOakSmolder\n",
       "2 M10_PeatSmolder \n",
       "3 M101_Saline     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read in metadata (subject info file)\n",
    "subjectinfo <- read_excel(\"SampleID_Information_102320.xlsx\")\n",
    "dim(subjectinfo) #170 rows, 7 col\n",
    "\n",
    "# Visualize this data quickly by viewing top left corner:\n",
    "subjectinfo[1:3,1:6]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Applying first round of various filters to these imported data files, including:\n",
    "- (1) First to first keep subjects present in both files\n",
    "- (2) Background filter, to filter out genes that were universally lowly expressed\n",
    "- (3) Check subject filter, to check that we have filtered out subjects that had samples present all values of zero (and thus not detected)\n",
    "- (4) Re-filter to include subjects meeting the above filters in (2) and (3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "##########\n",
    "# (1) Filtering to first keep subjects present in both files\n",
    "##########\n",
    "##########\n",
    "\n",
    "# Pull all mRNAseq IDs that overlap between countdata & subjectinfo\n",
    "keep_samples <- colnames(countdata[, colnames(countdata) %in% subjectinfo$FullID]) \n",
    "# Note that in this case, this represents 170 IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 3 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Gene</th><th scope=col>M1_RedOakSmolder</th><th scope=col>M10_PeatSmolder</th><th scope=col>M101_Saline</th><th scope=col>M107_RedOakFlame</th><th scope=col>M11_PeatSmolder</th><th scope=col>M113_PeatFlame</th><th scope=col>M119_PineNeedlesFlame</th><th scope=col>M12_PeatSmolder</th><th scope=col>M125_PineFlame</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Bcat1_29039 </td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Laptm5_29040</td><td>0</td><td>2</td><td>2</td><td>1</td><td>4</td><td>22</td><td>3</td><td>1</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Skp2_29041  </td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td> 0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & Gene & M1\\_RedOakSmolder & M10\\_PeatSmolder & M101\\_Saline & M107\\_RedOakFlame & M11\\_PeatSmolder & M113\\_PeatFlame & M119\\_PineNeedlesFlame & M12\\_PeatSmolder & M125\\_PineFlame\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & Bcat1\\_29039  & 0 & 0 & 0 & 0 & 0 &  0 & 0 & 0 & 0\\\\\n",
       "\t2 & Laptm5\\_29040 & 0 & 2 & 2 & 1 & 4 & 22 & 3 & 1 & 4\\\\\n",
       "\t3 & Skp2\\_29041   & 0 & 0 & 0 & 0 & 0 &  0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 10\n",
       "\n",
       "| <!--/--> | Gene &lt;chr&gt; | M1_RedOakSmolder &lt;dbl&gt; | M10_PeatSmolder &lt;dbl&gt; | M101_Saline &lt;dbl&gt; | M107_RedOakFlame &lt;dbl&gt; | M11_PeatSmolder &lt;dbl&gt; | M113_PeatFlame &lt;dbl&gt; | M119_PineNeedlesFlame &lt;dbl&gt; | M12_PeatSmolder &lt;dbl&gt; | M125_PineFlame &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Bcat1_29039  | 0 | 0 | 0 | 0 | 0 |  0 | 0 | 0 | 0 |\n",
       "| 2 | Laptm5_29040 | 0 | 2 | 2 | 1 | 4 | 22 | 3 | 1 | 4 |\n",
       "| 3 | Skp2_29041   | 0 | 0 | 0 | 0 | 0 |  0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Gene         M1_RedOakSmolder M10_PeatSmolder M101_Saline M107_RedOakFlame\n",
       "1 Bcat1_29039  0                0               0           0               \n",
       "2 Laptm5_29040 0                2               2           1               \n",
       "3 Skp2_29041   0                0               0           0               \n",
       "  M11_PeatSmolder M113_PeatFlame M119_PineNeedlesFlame M12_PeatSmolder\n",
       "1 0                0             0                     0              \n",
       "2 4               22             3                     1              \n",
       "3 0                0             0                     0              \n",
       "  M125_PineFlame\n",
       "1 0             \n",
       "2 4             \n",
       "3 0             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Filter the countdata dataframe for these IDs\n",
    "# Because the first column is the gene identifier information, this will get lost if we don't also grab that \n",
    "# separately (hence, a few steps here)\n",
    "countdata_temp <- countdata[, colnames(countdata) %in% keep_samples] # pulling countdata, without gene identifier col\n",
    "countdata <- cbind(countdata[,1], countdata_temp)    # adding the gene identifier column back in\n",
    "colnames(countdata)[1] <- \"Gene\"   # renaming the gene identifer column\n",
    "countdata_temp <- NULL    # removing temporary dataframe\n",
    "\n",
    "# Viewing merged, filtered count data\n",
    "countdata[1:3,1:10]\n",
    "\n",
    "# Filter the subjectinfo dataframe for these IDs\n",
    "subjectinfo <- subjectinfo[subjectinfo$FullID %in% keep_samples, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`summarise()` ungrouping output (override with `.groups` argument)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##########\n",
    "##########\n",
    "# (2) Background filter, to filter out genes that were universally lowly expressed\n",
    "##########\n",
    "##########\n",
    "\n",
    "\n",
    "# Create variable for number of samples, assuming that one variable is Gene: \n",
    "numberofsamples = (ncol(countdata)-1)\n",
    "\n",
    "# Create a variable that is the median of all expression levels across all samples and genes\n",
    "totalmedian = median(as.matrix(countdata[,2:ncol(countdata)]))\n",
    "\n",
    "# Create a list of genes that pass the background filter:\n",
    "Genes_backgroundfiltered <- countdata %>% \n",
    "  #gather so that there is one row for each gene-sample pairing\n",
    "  gather(key = \"sampleID\", value = \"expression\", 2:ncol(countdata)) %>% \n",
    "  #add a variable that takes the value 1 is the expression value is above the median \n",
    "  mutate(abovemedian = ifelse(expression > totalmedian,1,0)) %>% \n",
    "  #sum the abovemedian variable created above by gene so that totalabovemedian calcualtes the number of samples for each gene\n",
    "  #that have expression levels above the median\n",
    "  group_by(Gene) %>% \n",
    "  summarize(totalabovemedian = sum(abovemedian)) %>% \n",
    "  #remove all genes that have a totalabovemedian equal to less than a specified percentage of the total number of samples \n",
    "  filter(totalabovemedian > (0.20*numberofsamples)) %>% \n",
    "  #select only gene so that you essentially have a list of the genes to include \n",
    "  select(Gene) \n",
    "# Select genes that passed background filter only to proceed with analysis\n",
    "countdata <- left_join(Genes_backgroundfiltered, countdata, by = \"Gene\")\n",
    "# This resulted in filtering the original list of 30,146 genes down to 17,356 genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "##########\n",
    "#### (3) Check subject filter, to check that we have filtered out subjects that had samples present all values of zero\n",
    "#### (and thus not detected)\n",
    "##########\n",
    "##########\n",
    "\n",
    "# Note that this code was originally drafted using a transposed version of the countdata dataframe\n",
    "# So here, rather than modifying, we will transpose and then transpose back at the end\n",
    "\n",
    "#transpose count data \n",
    "countdata_t <- countdata %>% \n",
    "  gather(key = \"sampleID\", value = \"expression\", 2:ncol(countdata)) %>% \n",
    "  spread(Gene, expression)  \n",
    "# view transposed count data\n",
    "countdata_t[1:4,1:10]\n",
    "\n",
    "countdata_t <- countdata_t %>% \n",
    "  #create variable that is the sum of count values across all genes, for each sample\n",
    "  mutate(rowsum =  rowSums(countdata_t[,2:nrow(countdata_t)],na.rm  =  FALSE, dims = 1L)) \n",
    "\n",
    "#identify the samples that will be removed, for records \n",
    "samples_zerocounts <- countdata_t %>% \n",
    "  filter(rowsum  == 0) %>% \n",
    "  select(\"sampleID\") \n",
    "\n",
    "countdata_t <- countdata_t %>% \n",
    "  #remove samples that have a total sum of all counts of zero \n",
    "  filter(rowsum != 0) %>% \n",
    "  select(-rowsum)\n",
    "\n",
    "# SAMPLES REMOVED: No samples were removed \n",
    "\n",
    "\n",
    "# Transposing the countdata back such that the genes are the rownames and the samples are the columnnames\n",
    "countdata <- countdata_t %>% \n",
    "  gather(Gene, value, 2:ncol(countdata_t)) %>% \n",
    "  spread(sampleID, value) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "##########\n",
    "# (4) Re-filter to include subjects meeting the above filters in (2) and (3)\n",
    "##########\n",
    "##########\n",
    "\n",
    "# Pull all mRNAseq IDs that overlap between countdata & subjectinfo\n",
    "keep_samples <- colnames(countdata[, colnames(countdata) %in% subjectinfo$FullID]) \n",
    "# Note that in this case, this represents 372 IDs\n",
    "\n",
    "# Filter the countdata dataframe for these IDs\n",
    "# Because the first column is the gene identifier information, this will get lost if we don't also grab that separately (hence, a few steps here)\n",
    "countdata_temp <- countdata[, colnames(countdata) %in% keep_samples]   # pulling all countdata, without gene identifier column\n",
    "countdata <- cbind(countdata$Gene, countdata_temp)    # adding the gene identifier column back in\n",
    "colnames(countdata)[1] <- \"Gene\"   # renaming the gene identifer column\n",
    "# Viewing merged, filtered count data\n",
    "countdata[1:3,1:10]\n",
    "\n",
    "# Filter the subjectinfo dataframe for these IDs\n",
    "subjectinfo <- subjectinfo[subjectinfo$mRNAseq_ID %in% keep_samples, ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create dataframes that are formatted for proceeding code, as well as DESeq2 functions countdata and coldata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the gene variable the rowname \n",
    "countdata <- countdata %>% \n",
    "  column_to_rownames(var=\"Gene\") \n",
    "\n",
    "# Creating the coldata object, based on information in the subjectinfo file\n",
    "coldata <- subjectinfo\n",
    "# Let's change the name of the primary ID column, to use from here on out\n",
    "names(coldata)[names(coldata) == 'mRNAseq_ID'] <- 'ID'\n",
    "\n",
    "\n",
    "# Set the rownames of coldata and column names of countdata to be in the same order \n",
    "countdata <- setcolorder(countdata, as.character(coldata$ID))\n",
    "\n",
    "# Double checking that the same variables appear between the two dataframes\n",
    "setequal(as.character(coldata$ID), colnames(countdata))\n",
    "\n",
    "# Additionally checking that not only the sets of variables are the same, but that they are in the same order\n",
    "identical(as.character(coldata$ID), colnames(countdata))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNASeq QA/QC on raw count data to identify potential outlier samples. (This may or may not result in another filter step.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# One way to evaluate sequencing data quality / identify potential outliers is through Principal Component Analysis (PCA):\n",
    "# PCA helps in identifying outlying samples for quality control, and gives a feeling for the principal causes of variation in a dataset\n",
    "\n",
    "# Calculate principal components using transposed count data\n",
    "pca <- prcomp(t(countdata))\n",
    "screeplot(pca)\n",
    "#this scree plot indicates that nearly all variation is explained in PC1,2 and 3, \n",
    "#thus PCA is a useful exercise \n",
    "\n",
    "# Make dataframe for PCA plot generation using first three components\n",
    "pca_df <- data.frame(PC1 = pca$x[,1], PC2 = pca$x[,2], PC3 = pca$x[,3], Sample=colnames(countdata))\n",
    "# Add attributes (covariates from coldata) to pca df\n",
    "pca_df <- merge(pca_df, coldata, by.x=\"Sample\", by.y=\"ID\")\n",
    "\n",
    "\n",
    "\n",
    "# Calculating percent of the variation that is captured by each principal component\n",
    "pca_percent <- round(100*pca$sdev^2/sum(pca$sdev^2),1)\n",
    "\n",
    "# Generating PCA plot annotated by sex \n",
    "ggplot(pca_df, aes(PC1,PC2, color = sex))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Generating PCA plot annotated by maternal age \n",
    "ggplot(pca_df, aes(PC1,PC2, color = mage))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Generating PCA plot annotated by gestational age \n",
    "ggplot(pca_df, aes(PC1,PC2, color = gadays))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Generating PCA plot annotated by race \n",
    "ggplot(pca_df, aes(PC1,PC2, color = race1))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Lets identify the outliers\n",
    "ggplot(pca_df, aes(PC1,PC2))+\n",
    "  geom_text(aes(label=Sample, size=3)) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "\n",
    "# This would suggest that LS18017 and LS18097 are potential outliers, as well as potentially LS18183 and LS18076\n",
    "# We will conduct heirachical clustering to investigate further and to determine if we should also remove these samples\n",
    "\n",
    "\n",
    "# create a dataframe with samples as rows and genes as columns to input into hclust\n",
    "countdata_forclustering <- t(countdata)\n",
    "countdata_forclustering[1:5,1:10]\n",
    "\n",
    "\n",
    "# run heirarchical clustering\n",
    "hc <-hclust(dist(countdata_forclustering))\n",
    "jpeg(filename= paste0(Output_Folder,\"/\",cur_date, \"_heirachicalclusteringforoutliers.jpeg\"),width = 4500, height = 350)\n",
    "plot(hc)\n",
    "dev.off()\n",
    "\n",
    "# Here, the four aforementioned samples were clearly separate from the rest, so should be removed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Removing sample(s) with potential QA/QC issues & Rerunning PCA analysis with outlier removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove these subject IDs from our \"keep_samples\" vector\n",
    "keep_samples <- keep_samples[! keep_samples %in% c(\"LS18017\", \"LS18097\", \"LS18183\", \"LS18076\")]\n",
    "# Note that in this case, this represents 368 IDs\n",
    "\n",
    "# Filter the countdata dataframe for these IDs\n",
    "countdata <- countdata[, colnames(countdata) %in% keep_samples]\n",
    "countdata[1:5, 1:10]\n",
    "\n",
    "# Filter the coldata dataframe for these IDs\n",
    "coldata <- coldata[coldata$ID %in% keep_samples, ]\n",
    "\n",
    "\n",
    "\n",
    "# Export the raw data for just the included samples to potentially generate plots outside of R\n",
    "write.csv(countdata, paste0(Output_Folder,\"/\", cur_date, \"_RawCounts_SubjectsIncludedinModel.csv\"), row.names= TRUE)\n",
    "write.csv(coldata, paste0(Output_Folder,\"/\", cur_date, \"_SubjectInfo_SubjectsIncludedinModel.csv\"), row.names= FALSE)\n",
    "# Note that if, for future studies, we exclude subjects with missing covariate data, that step will have to incorporated prior to this export\n",
    "# But for this analysis, we're imputing missing covariate data, so went ahead and exported raw data here\n",
    "\n",
    "\n",
    "\n",
    "# Calculate principal components using transposed count data\n",
    "pca <- prcomp(t(countdata))\n",
    "screeplot(pca)\n",
    "#this scree plot indicates that nearly all variation is explained in PC1,2 and 3, \n",
    "#thus PCA is a useful exercise \n",
    "\n",
    "# Make dataframe for PCA plot generation using first three components\n",
    "pca_df <- data.frame(PC1 = pca$x[,1], PC2 = pca$x[,2], PC3 = pca$x[,3], Sample=colnames(countdata))\n",
    "# Add attributes (covariates from coldata) to pca df\n",
    "pca_df <- merge(pca_df, coldata, by.x=\"Sample\", by.y=\"ID\")\n",
    "\n",
    "# Calculating percent of the variation that is captured by each principal component\n",
    "pca_percent <- round(100*pca$sdev^2/sum(pca$sdev^2),1)\n",
    "\n",
    "# Generating PCA plot annotated by sex \n",
    "ggplot(pca_df, aes(PC1,PC2, color = sex))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Generating PCA plot annotated by maternal age \n",
    "ggplot(pca_df, aes(PC1,PC2, color = mage))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Generating PCA plot annotated by gestational age \n",
    "ggplot(pca_df, aes(PC1,PC2, color = gadays))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Generating PCA plot annotated by race \n",
    "ggplot(pca_df, aes(PC1,PC2, color = race1))+\n",
    "  geom_point(size=6) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))\n",
    "\n",
    "# Lets identify potential outliers (although here, since this is the 2nd run, we would be much stricter on this; meaning that there would have to be a super obvious outlier, or else the cycle would just continue)\n",
    "ggplot(pca_df, aes(PC1,PC2))+\n",
    "  geom_text(aes(label=Sample, size=3)) +\n",
    "  labs(x=paste0(\"PC1 (\",pca_percent[1],\"%)\"), y=paste0(\"PC2 (\",pca_percent[2],\"%)\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Setting up the actual DESeq2 function and associated algorithm (\"experiment\"). More information can be found about the DeSeq2 package [here](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# At this point, we need to make sure that the coldata is in the following format:\n",
    "# rownames = the subject ID that matches the countdata\n",
    "# all other columns indicate the covariate data that we may/may not include in the final experiment\n",
    "coldata[1:5, 1:7]   # viewing data\n",
    "\n",
    "\n",
    "# At this point, we need to make sure that the countdata is in the following format:\n",
    "# rownames = the gene identifiers\n",
    "# all other columns indicate individual samples that match (and are in the same order as) the coldata\n",
    "countdata[1:5, 1:10]   # viewing data\n",
    "\n",
    "\n",
    "# all values should be integers (DESeq2 requires integer count values), need to convert values here if haven't already\n",
    "sapply(countdata, class)   # checking the class of the values\n",
    "countdata_2 <- as.data.frame(sapply(countdata, as.integer))   # converting into integers if showing numeric\n",
    "sapply(countdata_2, class)   # re-checking the class of the values\n",
    "row.names(countdata_2) <- row.names(countdata)  # this removed the row names (gene identifiers), so had to merge these back\n",
    "countdata <- countdata_2\n",
    "\n",
    "\n",
    "\n",
    "# Create the final DESeq2 experiment, with appropriate experimental design:\n",
    "# Note in the design: last factor indicates the factor we want to test the effect of; the first factor(s) = factors we want to control for\n",
    "# Note that in this particular experiment, it gives us a warning that we are using integer data as numeric variables - this is what we want for some variables\n",
    "\n",
    "dds = DESeqDataSetFromMatrix(countData = countdata,\n",
    "                             colData = coldata,\n",
    "                             design = ~race1 + mage + sex + gadays + zcat4 + medu + ASD)\n",
    "\n",
    "# View what the experiment contains\n",
    "dds\n",
    "\n",
    "\n",
    "# Let's also make sure that we have the main contrast (last term) in the order we want to calculate appropriate fold change values\n",
    "# In this case, we want ASD=0 as denominators (first level, aka, the \"control\" level), and ASD=1, cases as numerators\n",
    "dds$ASD <- relevel (dds$ASD, \"0\")\n",
    "\n",
    "\n",
    "\n",
    "# Need to next, estimate the size factors, since size factors are used to normalize the counts (next step)\n",
    "# The \"iterate\" estimator iterates between estimating the dispersion with a design of ~1, and finding a size factor vector by numerically optimizing the likelihood of the ~1 model.\n",
    "dds <- estimateSizeFactors(dds)\n",
    "sizeFactors(dds) #check size factors\n",
    "\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "#### Pulling and exporting normalized count, for internal records and for potential future plots/figures\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "# Note that code is drafted to output several forms of normalized data - so users can choose which applies to their external purposes the best\n",
    "# Also note that variance stabilized (vsd) values are commonly used for figures, which requires the experiment to run and adjust for covariates below, so will export these in the code below\n",
    "\n",
    "\n",
    "# normalized counts:\n",
    "normcounts<- counts(dds, normalized=TRUE)\n",
    "write.csv(normcounts, paste0(Output_Folder,\"/\", cur_date, \"_NormCounts_IncludedSubjects.csv\"), row.names=TRUE)\n",
    "\n",
    "# pseudocounts to make plots easier:\n",
    "ps_normcounts <- normcounts + 1\n",
    "write.csv(ps_normcounts, paste0(Output_Folder,\"/\",cur_date, \"_NormCounts_ps_IncludedSubjects.csv\"),row.names=TRUE)\n",
    "\n",
    "# log2 pseudocounts (y=log2(n+1))\n",
    "log2normcounts <- log2(normcounts+1)\n",
    "write.csv(log2normcounts, paste0(Output_Folder,\"/\", cur_date, \"_NormCounts_pslog2_IncludedSubjects.csv\"), row.names=TRUE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "#### Surrogate variable analysis (SVA) to capture potential variances between samples (including batch, but also sample / cell type heterogeneity)\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "## For more information: https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.0030161\n",
    "## \"Capturing Heterogeneity in Gene Expression Studies by Surrogate Variable Analysis\", Jeffrey T Leek,  John D Storey\n",
    "\n",
    "\n",
    "# Set the model matrix for what's being used to fit the data (minus the SVA variables)\n",
    "mod <- model.matrix(~race1 + mage + sex + gadays + zcat4 + medu + ASD, colData(dds))\n",
    "\n",
    "\n",
    "# Set the null model matrix being compared against when fitting the data for the SVA analysis\n",
    "mod0 <- model.matrix( ~1, colData(dds))\n",
    "\n",
    "\n",
    "# Running SVA, here, using number of SVs = 3\n",
    "# Note that this number of significant SVs selection was based on several iterations, where different numbers of SVs were tested for the ELGAN dataset\n",
    "svseq <- svaseq(normcounts, mod, mod0, n.sv=3)\n",
    "\n",
    "# I also tested an iteration where svaseq calculated the number of significant SVs, using the following code:\n",
    "# svseq <- svaseq(normcounts, mod, mod0, n.sv=NULL)  \n",
    "# Number of significant surrogate variables is much higher, but this muted global signal responses\n",
    "\n",
    "\n",
    "#################\n",
    "#################\n",
    "#### Adding surrogate variables into the DESeq2 analysis\n",
    "#################\n",
    "#################\n",
    "\n",
    "dds$SV1 <- svseq$sv[,1]\n",
    "dds$SV2 <- svseq$sv[,2]\n",
    "dds$SV3 <- svseq$sv[,3]\n",
    "design(dds) <- ~ SV1 + SV2 + SV3 + race1 + mage + sex + gadays + zcat4 + medu + ASD\n",
    "\n",
    "\n",
    "# Export variance stabilized counts, which will be influenced by the new design above\n",
    "# vsd will be useful for plots and future figure generation\n",
    "# CAUTION: THIS STEP TAKES SOME TIME! RUN TIME IS ABOUT 20 MINUTES FOR MRNASEQ DATA ON JULIA'S COMPUTER\n",
    "vsd <- varianceStabilizingTransformation(dds, blind=FALSE)\n",
    "vsd_matrix <-as.matrix(assay(vsd))\n",
    "write.csv(vsd_matrix, paste0(Output_Folder,\"/\", cur_date, \"_VSDCounts_IncludedSubjects.csv\"), row.names=TRUE)\n",
    "\n",
    "\n",
    "\n",
    "#################\n",
    "#################\n",
    "#### Evaluating variance stabilized values to guage whether SVs accounted for unwanted sources of variation between samples\n",
    "#################\n",
    "#################\n",
    "\n",
    "plotPCA(vsd, intgroup=\"sex\", returnData=FALSE)\n",
    "plotPCA(vsd, intgroup=\"mage\", returnData=FALSE)\n",
    "plotPCA(vsd, intgroup=\"SV1\", returnData=FALSE)\n",
    "#SV1 accounts for a lot of variation along PC1 as shown by the steady gradient in color across the PC1 axis\n",
    "plotPCA(vsd, intgroup=\"SV2\", returnData=FALSE)\n",
    "plotPCA(vsd, intgroup=\"SV3\", returnData=FALSE)\n",
    "#SV2 and 3 also account for a lot of variation across PCs\n",
    "\n",
    "#In sum, it looks like the SVs are accounting for a lot of the unwanted variation which is great!\n",
    "\n",
    "\n",
    "\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "#### Statistical analysis to detect significantly differentially expressed genes\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "\n",
    "# Running the differential expression statistical pipeline\n",
    "# CAUTION: THIS STEP TAKES A LOT OF TIME! RUN TIME IS ABOUT 1 HOUR FOR MRNASEQ DATA ON JULIA'S COMPUTER\n",
    "dds <- DESeq(dds, betaPrior=FALSE)      # because we used a user-defined model matrix, need to set betaPrior=FALSE\n",
    "# This ran a Wald test p-value (ASD 1 vs. 0)\n",
    "\n",
    "resultsNames(dds) #check the available comparisons\n",
    "\n",
    "# Pulling statistical results\n",
    "res <- results(dds, pAdjustMethod = \"BH\")  #The  with multiple test correction by the default, BH (aka FDR)\n",
    "head(res)\n",
    "\n",
    "\n",
    "# Exporting statistical results:\n",
    "write.csv(as.data.frame(res)[order(res$padj),], paste0(Output_Folder,\"/\", cur_date,\"_AllStatResults.csv\"))\n",
    "\n",
    "\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "#### Overview of statistical results\n",
    "#################################################################################################\n",
    "#################################################################################################\n",
    "summary(res)\n",
    "\n",
    "# Basic filters\n",
    "table(res$padj<0.10)\n",
    "\n",
    "table(res$padj<0.10 & res$log2FoldChange > log2(1.5))\n",
    "table(res$padj<0.10 & res$log2FoldChange < -log2(1.5))\n",
    "\n",
    "table(res$padj<0.10 & res$log2FoldChange > log2(1.3))\n",
    "table(res$padj<0.10 & res$log2FoldChange < -log2(1.3))\n",
    "\n",
    "table(res$padj<0.10 & (res$log2FoldChange > log2(1.5) | res$log2FoldChange < -log2(1.5)))\n",
    "\n",
    "\n",
    "# Sort results and create a dataframe (df) for exporting & certain plots in the code below\n",
    "resdf <- as.data.frame(res)[order(res$padj),]\n",
    "\n",
    "\n",
    "\n",
    "# Also create a df for filtered data for certain plots in the code below\n",
    "resdf_filter <- resdf[which(resdf$padj<0.10 & (resdf$log2FoldChange > log2(1.5) | resdf$log2FoldChange < -log2(1.5))),]\n",
    "resdf_pfilter <- resdf[which(resdf$padj<0.10),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Heatmap###\n",
    "\n",
    "#make dataset for heatmap in partek\n",
    "countsforhm <- vsd_matrix\n",
    "\n",
    "resdf_pfilter <- resdf_pfilter %>% \n",
    "  rownames_to_column(\"gene\") %>% \n",
    "  filter(abs(log2FoldChange)>log2(1.3)) %>% \n",
    "  arrange(log2FoldChange)\n",
    "\n",
    "genenames <- as.data.frame(resdf_pfilter$gene) %>% \n",
    "  rename(\"gene\"=\"resdf_pfilter$gene\")\n",
    "\n",
    "countsforhm <- as.data.frame(countsforhm) %>% \n",
    "  rownames_to_column(var=\"gene\") \n",
    "\n",
    "countsforhm <- left_join(genenames, countsforhm, by=\"gene\") %>% \n",
    "  column_to_rownames(var=\"gene\")\n",
    "\n",
    "countsforhm <- as.data.frame(t(countsforhm)) %>% \n",
    "  rownames_to_column(var=\"ID\") \n",
    "\n",
    "countsforhm <- full_join(countsforhm, coldata, by=\"ID\") %>% \n",
    "  select(-\"race\",-\"gadays\",- \"mage\")\n",
    "\n",
    "countsforhm <- countsforhm %>% \n",
    "  column_to_rownames(var=\"ID\")\n",
    "\n",
    "write.csv(countsforhm, paste0(Output_Folder,\"/\", cur_date,\"_countdata_forheatmap_mrnas.csv\"))\n",
    "\n",
    "### MA plot ###\n",
    "\n",
    "###################################\n",
    "####  MA plot in ggplot2\n",
    "###################################\n",
    "MA <- resdf\n",
    "MA[is.na(MA)] <- 1\n",
    "MA1 <- MA[ which(MA$padj>=0.1),]  # all the rest\n",
    "MA2 <- MA[ which(MA$padj<0.1 & MA$log2FoldChange > log2(1.3)),]   # sig up-regulated w/ FC filter-FIREBRICK\n",
    "MA3 <- MA[ which(MA$padj<0.1 & MA$log2FoldChange < -log2(1.3)),]  # sig down-regulated w/ FC filter-dodgerblue4\n",
    "MA4 <- MA[ which(MA$padj<0.1 & MA$log2FoldChange > 0 & MA$log2FoldChange <= log2(1.3)),]  # sig up-regulated w/o FC filter- darkorchid3\n",
    "MA5 <- MA[ which(MA$padj<0.1 & MA$log2FoldChange < 0 & MA$log2FoldChange >= -log2(1.3)), ] # sig down-regulated w/o FC filter- chartreuse4\n",
    "\n",
    "# get names of probes if annotating the graphic with them, according to specific criteria (not used here):\n",
    "namesmmu=rownames(MA)[which((MA$log2FoldChange > 2 | MA$log2FoldChange < -2) & MA$padj < 0.1 & MA$baseMean > 0)]\n",
    "namesmmu2=rownames(MA2)[which((MA2$log2FoldChange > 2 | MA2$log2FoldChange < -2) & MA2$padj < 0.1 & MA2$padj >= 0.05 & MA2$baseMean > 0)]\n",
    "namesmmu3=rownames(MA3)[which((MA3$log2FoldChange > 2 | MA3$log2FoldChange < -2) & MA3$padj < 0.05 & MA3$baseMean > 0)]                                               \n",
    "\n",
    "plot <-\n",
    "  ggplot(subset(MA1, baseMean>=0), aes(x = baseMean, y = log2FoldChange)) + #can subset to exclude the very low counts (<1 average count) if we want\n",
    "  geom_point(color=\"gray56\", size = 2) + \n",
    "  geom_point(data = MA2, color=\"firebrick\", size=3, show.legend = TRUE) + # colors used: firebrick, dodgerblue4, chocolate1, darkorchid3, chartreuse4\n",
    "  geom_point(data = MA3, color=\"dodgerblue4\", size=3, show.legend = TRUE) +\n",
    "  geom_point(data = MA4, color=\"darkorchid3\", size=3, show.legend = TRUE) +\n",
    "  geom_point(data = MA5, color=\"chartreuse4\", size=3, show.legend = TRUE) +\n",
    "  theme_bw() +\n",
    "  scale_x_continuous(trans = \"log10\", breaks=c(1,10,100, 1000, 10000, 100000, 1000000), labels=c(\"1\",\"10\",\"100\", \"1000\", \"10000\", \"100000\", \"1000000\")) + \n",
    "  scale_y_continuous(limits=c(-1.5, 1.5), breaks=c(1.5,1,0.5,0,-0.5,-1,-1.5), labels=c(1.5,1,0.5,0,-0.5,-1,-1.5)) +\n",
    "  xlab(\"Expression (Normalized Count)\") + ylab(\"Fold Change (log2) Cases/Control\") +\n",
    "  labs(title=\"ASD vs. Control\") +   \n",
    "  #coord_fixed(ratio=0.5) +\n",
    "  geom_hline(yintercept=0) \n",
    "#annotate(\"text\",label=\"# DEPs higher in Cr(VI): 78\\n(representing 77 DEGs)\", x=1.1,y=10, size=4, hjust=0, color=\"darkorchid3\") +\n",
    "#annotate(\"text\",label=\"# DEPs higher in Captan: 86\\n(representing 85 DEGs)\", x=1.1,y=-10, size=4, hjust=0, color=\"chartreuse4\")+\n",
    "#geom_text(data=MA2[namesmmu2,], size=3, hjust=-.05, vjust=.7, col=\"deepskyblue\", aes(label=namesmmu2,size=5, angle=0)) + #annotate specific probes with text\n",
    "#geom_text(data=MA3[namesmmu3,], size=3, hjust=-.05, vjust=.7, col=\"firebrick3\", aes(label=namesmmu3,size=5, angle=0))\n",
    "\n",
    "plot\n",
    "\n",
    "#export single plots\n",
    "png(paste0(Output_Folder, \"/\", cur_date, \"_MAplot_v1.png\"),\n",
    "    width = 1600, height = 1200, units = \"px\", res = 250)\n",
    "plot\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
